### PR TITLE
add kafka adapter

### PR DESCRIPTION
**Pull request by LinuxGit**

### CODE REVIEW

File: kafka-adapter/httpapi.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+// Alerts is a list of Alert objects.
+type Alerts []Alert
+
+// KV is a set of key/value string pairs.
+type KV map[string]string
+
+//AlertMsgFromWebhook get alert massage
+func (r *Run) AlertMsgFromWebhook(w http.ResponseWriter, hr *http.Request) {
+	b, err := ioutil.ReadAll(hr.Body)
```
Comment (zhouqiang-cl) [inline kafka-adapter/httpapi.go]: extract read function

File: kafka-adapter/kafka-adapter.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+func initLog() error {
+	log.SetLevelByString(*logLevel)
+	if *logFile != "" {
+		return log.SetOutputByName(*logFile)
+	}
+	return nil
+}
+
+//Scheduler for monitor chann data
```
Comment (zhouqiang-cl) [inline kafka-adapter/kafka-adapter.go]: how to close scheduleer

File: kafka-adapter/kafka-adapter.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	URL         string `json:"url"`
+	Level       string `json:"level"`
+	Note        string `json:"note"`
+	Value       string `json:"value"`
+	Time        string `json:"time"`
+}
+
+//Run represent runtime informations
+type Run struct {
+	Rdr         *render.Render
```
Comment (zhouqiang-cl) [inline kafka-adapter/kafka-adapter.go]: struct Run defined in kafka-adapter.go and interface also in kafka.go, uniform it

File: kafka-adapter/kafka.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	kafkaMsg := &sarama.ProducerMessage{
+		Topic: *kafkaTopic,
+		Value: sarama.StringEncoder(msg),
+	}
+	log.Debugf("get kafka mssage %s", msg)
+	_, _, err := r.KafkaClient.SendMessage(kafkaMsg)
+	return err
+}
+
+//TransferData transfer alert to kafka string
```
Comment (zhouqiang-cl) [inline kafka-adapter/kafka.go]: If error, what happends

Comment (zhouqiang-cl) [reply kafka-adapter/kafka.go]: you can return an error in transferData

File: kafka-adapter/kafka.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package main
+
+import (
+	//"fmt"
```
Comment (zhouqiang-cl) [inline kafka-adapter/kafka.go]: delete this line

File: kafka-adapter/httpapi.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/ngaut/log"
+	"github.com/unrolled/render"
+)
+
+// AlertData alertmanager base struct
```
Comment (GregoryIan) [inline kafka-adapter/httpapi.go]: AlertData is base structure of alert manager

File: kafka-adapter/httpapi.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	log.Debugf("get alert data b %v", string(b))
+	defer hr.Body.Close()
+	alertData := &AlertData{}
+	err = json.Unmarshal(b, alertData)
+	if err != nil {
+		log.Errorf("can not unmarshal http post data with error %v", err)
+		r.Rdr.Text(w, http.StatusBadRequest, err.Error())
+		return
+	}
+	log.Debugf("get alert data %v", alertData)
```
Comment (GregoryIan) [inline kafka-adapter/httpapi.go]: log.Infof("alert data %+v", alertData)

File: kafka-adapter/kafka-adapter.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/ngaut/log"
+	"github.com/unrolled/render"
+)
+
+var (
+	port         = flag.Int("port", 28082, "port to listen on for the web interface")
+	kafkaAddress = flag.String("kafka-address", "10.0.3.4:9092,10.0.3.5:9092,10.0.3.6:9092", "kafka address")
```
Comment (GregoryIan) [inline kafka-adapter/kafka-adapter.go]: "" and verify it to make sure it's not empty

File: kafka-adapter/kafka-adapter.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	kafkaTopic   = flag.String("kafka-topic", "test", "kafka topic")
```
Comment (GregoryIan) [inline kafka-adapter/kafka-adapter.go]: don't use test default

File: kafka-adapter/kafka.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	}
+}
+
+//Scheduler for monitoring chan data
+func (r *Run) Scheduler() {
+	for {
+		lenAlertMsgs := len(r.AlertMsgs)
+		if lenAlertMsgs > 0 {
+			for i := 0; i < lenAlertMsgs; i++ {
+				r.TransferData(<-r.AlertMsgs)
```
Comment (zhouqiang-cl) [inline kafka-adapter/kafka.go]: alertMsgs is an array

File: kafka-adapter/kafka.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+}
+
+//Scheduler for monitoring chan data
+func (r *Run) Scheduler() {
+	for {
+		lenAlertMsgs := len(r.AlertMsgs)
+		if lenAlertMsgs > 0 {
+			for i := 0; i < lenAlertMsgs; i++ {
```
Comment (zhouqiang-cl) [inline kafka-adapter/kafka.go]: using range?

File: kafka-adapter/main.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+		syscall.SIGKILL,
+		syscall.SIGHUP,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+
+	go func() {
+		sig := <-sc
+		log.Infof("got signal [%d] to exit.", sig)
+		r.KafkaClient.Close()
```
Comment (zhouqiang-cl) [inline kafka-adapter/main.go]: close schedualer

File: kafka-adapter/httpapi.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		log.Errorf("can not unmarshal http post data with error %v", err)
+		r.Rdr.Text(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	log.Infof("alert data %+v", alertData)
+
+	r.AlertMsgs <- alertData
+	r.Rdr.Text(w, http.StatusAccepted, "")
```
Comment (GregoryIan) [inline kafka-adapter/httpapi.go]: `r.Rdr.Text(w, http.StatusAccepted, "success")`

File: kafka-adapter/kafka.go [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+	if val, ok := kv[key]; ok {
+		return val
+	}
+	return ""
+}
+
+//CreateKafkaProducer creates a new SyncProducer using the given broker addresses and configuration
+func (r *Run) CreateKafkaProducer() error {
+	config := sarama.NewConfig()
+	config.Producer.RequiredAcks = sarama.WaitForLocal
```
Comment (GregoryIan) [inline kafka-adapter/kafka.go]: it's default config. and u can lose some data, if partition is down

**Comment by GregoryIan** [comment]
it can wrok now, LGTM. but u need to refine it later and read kafka client doc

**Pull request merged**